File: 2021_random_set_self_practice/using_python/탈옥수.py
"""
input :
5
0.000 0.000
1.000 1.000
2.000 0.000
3.000 1.500
2.235 2.483

output :
RIGHT
LEFT
LEFT
"""
import sys

sys.stdin = open("input.txt", "r")
input = sys.stdin.readline


class Prisoner(object):
    def __init__(self):
        self.N = -1
        self.seq = []

    def input(self):
        self.N = int(input())
        for _ in range(self.N):
            x, y = map(float, input().rstrip().split())
            self.seq.append((x, y))

    def ccw(self, p1, p2, p3):
        ret = (p2[0] - p1[0]) * (p3[1] - p2[1]) - (p2[1] - p1[1]) * (p3[0] - p2[0])
        if ret > 0:
            return 1
        elif ret < 0:
            return -1
        else:
            return 0

    def go(self):
        idx = 0
        while idx != len(self.seq) - 2:
            ret = self.ccw(self.seq[idx], self.seq[idx + 1], self.seq[idx + 2])
            if ret == 1:
                print("LEFT")
            elif ret == -1:
                print("RIGHT")
            else:
                print("error")
            idx += 1


if __name__ == "__main__":
    prisoner = Prisoner()
    prisoner.input()
    prisoner.go()

File: 2021_random_set_self_practice/using_cpp/탈옥수.cpp
/*
input :
5
0.000 0.000
1.000 1.000
2.000 0.000
3.000 1.500
2.235 2.483

output :
RIGHT
LEFT
LEFT
 */
#include <iostream>
#define SZ (10003)

using namespace std;

typedef struct coordinate {
    double x;
    double y;
}coordinate;


class Prisoner {
public:
    int N;
    coordinate seq[SZ];

    void input();
    int ccw(coordinate p1, coordinate p2, coordinate p3);
    void go();
};

void Prisoner::input() {
    // input
    freopen("input.txt", "r", stdin);
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> this->N;
    double a;
    double b;
    for (int i = 0; i < this->N; i++) {
        cin >> a;
        cin >> b;
        seq[i] = {a, b};
    }
}

int Prisoner::ccw(coordinate p1, coordinate p2, coordinate p3) {
    double ret = (p2.x - p1.x) * (p3.y - p2.y) - (p2.y - p1.y) * (p3.x - p2.x);
    if (ret > 0) {
        return 1;
    }
    else if (ret < 0) {
        return -1;
    }
    else{
        return 0;
    }
}

void Prisoner::go() {
    int idx = 0;
    while (idx != this->N - 2) {
        int ret = ccw(seq[idx], seq[idx + 1], seq[idx + 2]);
        switch (ret) {
            case 1:
                cout << "LEFT\n";
                break;
            case -1:
                cout << "RIGHT\n";
                break;
            default:
                /* invalid entry */
                break;
        }
        idx++;
    }
}

int main() {
    Prisoner prisoner;
    prisoner.input();
    prisoner.go();

    return 0;
}